src/rowbinary/de.rs

```diff
@@ -242,42 +400,204 @@ impl<'data> Deserializer<'data> for &mut RowBinaryDeserializer<'_, 'data> {
+            }
+        }
+
         let len = self.read_size()?;
-        self.deserialize_tuple(len, visitor)
+        let inner_data_type_validator =
+            self.validator
+                .validate(&SerdeType::Seq, &[DataTypeHint::Array], len)?;
+        visitor.visit_seq(RowBinarySeqAccess {
+            deserializer: &mut RowBinaryDeserializer {
+                input: self.input,
+                validator: inner_data_type_validator,
+            },
+            len,
+        })
     }
 
     #[inline]
-    fn deserialize_map<V: Visitor<'data>>(self, _visitor: V) -> Result<V::Value> {
-        panic!("maps are unsupported, use `Vec<(A, B)>` instead");
+    fn deserialize_map<V: Visitor<'data>>(self, visitor: V) -> Result<V::Value> {
+        // println!(
+        //     "deserialize_map call",
+        // );
+
+        struct RowBinaryMapAccess<'de, 'cursor, 'data, Validator>
+        where
+            Validator: ValidateDataType,
+        {
+            deserializer: &'de mut RowBinaryDeserializer<'cursor, 'data, Validator>,
+            entries_visited: usize,
+            len: usize,
+        }
+
+        impl<'data, Validator> MapAccess<'data> for RowBinaryMapAccess<'_, '_, 'data, Validator>
+        where
+            Validator: ValidateDataType,
+        {
+            type Error = Error;
+
+            fn next_key_seed<K>(&mut self, seed: K) -> Result<Option<K::Value>>
+            where
+                K: DeserializeSeed<'data>,
+            {
+                if self.entries_visited >= self.len {
+                    return Ok(None);
+                }
+                self.entries_visited += 1;
+                seed.deserialize(&mut *self.deserializer).map(Some)
+            }
+
+            fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value>
+            where
+                V: DeserializeSeed<'data>,
+            {
+                seed.deserialize(&mut *self.deserializer)
+            }
+
+            fn size_hint(&self) -> Option<usize> {
+                Some(self.len)
+            }
+        }
+
+        let len = self.read_size()?;
+        let inner_data_type_validator =
+            self.validator
+                .validate(&SerdeType::Map, &[DataTypeHint::Map], len)?;
+        visitor.visit_map(RowBinaryMapAccess {
+            deserializer: &mut RowBinaryDeserializer {
+                input: self.input,
+                validator: inner_data_type_validator,
+            },
+            entries_visited: 0,
+            len,
+        })
     }
 
     #[inline]
     fn deserialize_struct<V: Visitor<'data>>(
         self,
-        _name: &str,
+        name: &str,
         fields: &'static [&'static str],
         visitor: V,
     ) -> Result<V::Value> {
-        self.deserialize_tuple(fields.len(), visitor)
+        // println!("deserialize_struct: {} (fields: {:?})", name, fields,);
+
+        // FIXME use &'_ str, fix lifetimes
+        self.validator.set_struct_name(name.to_string());
```

Comment:
This is necessary for pretty error messages that include struct name... but I could not quickly find a proper way to pass it by reference.